cosmwasm/packages/wasmi-runtime/src/wasm/contract_operations.rs

```diff
@@ -67,34 +68,44 @@ pub fn init(
         .write_to_memory(&secret_msg.decrypt()?)
         .map_err(wasmi_error_to_enclave_error)?;
 
-    let vec_ptr = engine
-        .init(env_ptr, msg_ptr)
-        .map_err(wasmi_error_to_enclave_error)?;
-
-    let output = engine
-        .extract_vector(vec_ptr)
-        .map_err(wasmi_error_to_enclave_error)?;
-
-    trace!("Init output before encryption: {:?}", output);
-
-    // TODO: copy cosmwasm's structures to enclave
-    // TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/init_handle.rs#L129
-    // TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/query.rs#L13
-    let output = encrypt_output(output, secret_msg.nonce, secret_msg.user_public_key)?;
-
-    trace!("Init output after encryption: {:?}", output);
+    // This wrapper is used to coalesce all errors in this block to one object
+    // so we can `.map_err()` in one place for all of them
+    let mut wrapper = || -> Result<_, EnclaveError> {
+        let vec_ptr = engine
+            .init(env_ptr, msg_ptr)
+            .map_err(wasmi_error_to_enclave_error)?;
+
+        let output = engine
+            .extract_vector(vec_ptr)
+            .map_err(wasmi_error_to_enclave_error)?;
+
+        trace!("Init output before encryption: {:?}", output);
+
+        // TODO: copy cosmwasm's structures to enclave
+        // TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/init_handle.rs#L129
+        // TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/query.rs#L13
+        let output = encrypt_output(output, secret_msg.nonce, secret_msg.user_public_key)?;
+
+        trace!("Init output after encryption: {:?}", output);
+        Ok(output)
+    };
+    let output = wrapper().map_err(|err| {
+        *used_gas = engine.gas_used();
+        err
+    })?;
 
+    *used_gas = engine.gas_used();
     // todo: can move the key to somewhere in the output message if we want
     Ok(InitSuccess {
         output,
-        used_gas: engine.gas_used(),
         signature: contract_key,
     })
 }
```

Comment:
Ok so I hate this. I did it to not repeat this code:
```rust
.map_err(|err| {
    *used_gas = engine.gas_used();
    err
})
```
which has to be performed on each fallible operation from the call into the module until the return from this function.
We have two alternatives to this:

### Local variable and manually `map_err` for each fallible expression

In which case the code will look like this:
```rust
let report_gas = |err| {
    *used_gas = engine.gas_used();
    err
};

let vec_ptr = engine
    .init(env_ptr, msg_ptr)
    .map_err(wasmi_error_to_enclave_error)
    .map_err(report_gas)?;

let output = engine
    .extract_vector(vec_ptr)
    .map_err(wasmi_error_to_enclave_error)
    .map_err(report_gas)?;

trace!("Init output before encryption: {:?}", output);

// TODO: copy cosmwasm's structures to enclave
// TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/init_handle.rs#L129
// TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/query.rs#L13
let output =
    encrypt_output(output, secret_msg.nonce, secret_msg.user_public_key).map_err(report_gas)?;

trace!("Init output after encryption: {:?}", output);

*used_gas = engine.gas_used();
// todo: can move the key to somewhere in the output message if we want
Ok(InitSuccess {
    output,
    signature: contract_key,
})
```

### wrap the pattern into a macro

In which case the code will look like this:
```rust
// This at the top of the file or even in `lib.rs`
macro_rules! coalesce_err {
    ($error_type: ty, $body: expr) => {{
        #[allow(unused_mut)]
        let mut wrapper = || -> std::result::Result<_, $error_type> { Ok($body) };
        wrapper()
    }};
}

// and in the function:

// This wrapper is used to coalesce all errors in this block to one object
// so we can `.map_err()` in one place for all of them
let output = coalesce_err!(EnclaveError, {
    let vec_ptr = engine
        .init(env_ptr, msg_ptr)
        .map_err(wasmi_error_to_enclave_error)?;

    let output = engine
        .extract_vector(vec_ptr)
        .map_err(wasmi_error_to_enclave_error)?;

    trace!("Init output before encryption: {:?}", output);

    // TODO: copy cosmwasm's structures to enclave
    // TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/init_handle.rs#L129
    // TODO: ref: https://github.com/CosmWasm/cosmwasm/blob/b971c037a773bf6a5f5d08a88485113d9b9e8e7b/packages/std/src/query.rs#L13
    let output = encrypt_output(output, secret_msg.nonce, secret_msg.user_public_key)?;

    trace!("Init output after encryption: {:?}", output);
    output
})
.map_err(|err| {
    *used_gas = engine.gas_used();
    err
})?;

*used_gas = engine.gas_used();
// todo: can move the key to somewhere in the output message if we want
Ok(InitSuccess {
    output,
    signature: contract_key,
})
```

The technique used in the PR, and the second alternative both allow marking a region in which the gas will always be reported. We can even move the free `*used_gas = engine.gas_used();` line inside the marked block. The macro option is even nicer in my opinion because it does `Ok`-wrapping, but in both cases a `return` applied to the block, not the whole function.
The first alternative is a lot simpler, but error prone if someone adds a `?` expression without `.map_err(report_gas)`.

Reply:
We used `map_err` + `?` everywhere else in the code, but I don't really have a strong opinion on this.
I think in the first case the compiler won't let you use `let` without `?`.

Reply:
I'm fine with the macro, as long it's properly documented for those of us who aren't super familiar with macro syntax. Also, as long as we don't think that we will need to handle multiple error types (i.e. in the future we may want to return not only `EnclaveError`) and that it will be reused in other parts of the code.

Question - if we're adding the gas used every time we error out, why not return it as a part of the function call itself, rather than adding it after the fact? I mean change handle()/init/etc. to return Result<u32, (InterpreterError, GasUsed)> [or even create a new error type that contains both the error and the gas]
Another (unrelated to this PR) question: if we're mapping the error from handle()/etc/ to wasmi error anyway why aren't we returning EnclaveError in the first place?

Reply:
As discussed - I'll switch this out to use the macro. Cleaning up the error handling will come soon after but as a separate PR.